src/librustdoc/clean/utils.rs

```diff
@@ -73,31 +80,128 @@ pub(crate) fn krate(cx: &mut DocContext<'_>) -> Crate {
 
 pub(crate) fn substs_to_args<'tcx>(
     cx: &mut DocContext<'tcx>,
-    substs: ty::Binder<'tcx, &[ty::subst::GenericArg<'tcx>]>,
-    mut skip_first: bool,
+    substs: ty::Binder<'tcx, &[ty::GenericArg<'tcx>]>,
+    skip_first: bool,
+    owner: Option<DefId>,
 ) -> Vec<GenericArg> {
-    let mut ret_val =
-        Vec::with_capacity(substs.skip_binder().len().saturating_sub(if skip_first {
-            1
-        } else {
-            0
-        }));
-    ret_val.extend(substs.iter().filter_map(|kind| match kind.skip_binder().unpack() {
-        GenericArgKind::Lifetime(lt) => {
-            Some(GenericArg::Lifetime(clean_middle_region(lt).unwrap_or(Lifetime::elided())))
-        }
-        GenericArgKind::Type(_) if skip_first => {
-            skip_first = false;
-            None
-        }
-        GenericArgKind::Type(ty) => {
-            Some(GenericArg::Type(clean_middle_ty(kind.rebind(ty), cx, None)))
-        }
-        GenericArgKind::Const(ct) => {
-            Some(GenericArg::Const(Box::new(clean_middle_const(kind.rebind(ct), cx))))
+    // An empty ParamEnv is probably not the best choice but the simplest one and it should only
+    // lead to false negatives wrt. elision which should be fine. FIXME: Is this correct?
+    let param_env = ty::ParamEnv::empty();
+    let cause = ObligationCause::dummy();
+    let params = owner.map(|owner| &cx.tcx.generics_of(owner).params);
+    let mut elision_has_failed_once_before = false;
+
+    let offset = if skip_first { 1 } else { 0 };
+    let mut args = Vec::with_capacity(substs.skip_binder().len().saturating_sub(offset));
+
+    let subst_to_arg = |(index, arg): (usize, &ty::GenericArg<'tcx>)| {
```

Comment:
Can you refactor this closure to a nested function definition? If it doesn't actually close on its environment, it could be a bit better for perf